lessons/insertion.md

```diff
@@ -5,3 +5,154 @@ order: "5E"
+   int holePosition
+   int valueToInsert
+	
+   for i = 1 to length(A) inclusive do:
+	
+      /* select value to be inserted */
+      valueToInsert = A[i]
+      holePosition = i
+      
+      /*locate hole position for the element to be inserted */
+		
+      while holePosition > 0 and A[holePosition-1] > valueToInsert do:
+         A[holePosition] = A[holePosition-1]
+         holePosition = holePosition -1
+      end while
+		
+      /* insert the number at hole position */
+      A[holePosition] = valueToInsert
+      
+   end for
+	
+end procedure
+```
+
+
+
+## Complexity analysis
+
+Insertion sort runs in *O*(*n*) time in its best case and runs in *O*(*n*²) in its worst and average cases.
+
+### Best case
+
+Insertion sort performs two operations: it scans through the list, comparing each pair of elements, and it swaps elements if they are out of order. Each operation contributes to the running time of the algorithm. If the input array is already in sorted order, insertion sort compares *O*(*n*) elements and performs no swaps. Therefore, in the best case, insertion sort runs in O(*n*) time.
+
+### Worst case
+
+The worst case for insertion sort will occur when the input list is in decreasing order. To insert the last element, we need at most *n*−1 comparisons and at most *n*−1 swaps. To insert the second to last element, we need at most *n*−2 comparisons and at most *n*−2 swaps, and so on. The number of operations needed to perform insertion sort is therefore: 2 × (1+2+⋯+*n* −2+*n* −1) . This is in the order of sum of first n-1 numbers which is quadratic in nature. therefore, in the worst case, insertion sort runs in O(*n*²) time.
+
+
+
+> The average case and the worst case have the same time complexity. Try to think why is this happening.
+
+
+
+### Code Implementation 
+
+```java
+
+// Java program for implementation of Insertion Sort
+class InsertionSort {
+    /*Function to sort array using insertion sort*/
+    void insSort(int arr[])
+    {
+        int n = arr.length;
+        for (int i = 1; i < n; ++i) {
+            int key = arr[i];
+            int j = i - 1;
+ 
+            /* Move elements of arr[0..i-1], that are
+               greater than key, to one position ahead
+               of their current position */
+            while (j >= 0 && arr[j] > key) {
+                arr[j + 1] = arr[j];
+                j = j - 1;
+            }
+            arr[j + 1] = key;
+        }
+    }
+ 
+    /* A utility function to print array of size n*/
+    static void printArray(int arr[])
+    {
+        int n = arr.length;
+        for (int i = 0; i < n; ++i)
+            System.out.print(arr[i] + " ");
+ 
+        System.out.println();
+    }
+ 
+    public static void main(String args[])
+    {
+        int arr[] = { 1,5,4,2,3 };
+ 
+        InsertionSort ob = new InsertionSort();
+        ob.insSort(arr);
+ 
+        printArray(arr);
+    }
+} /* This code is contributed by Abhijit Mishra. */
```

Comment:
```suggestion
public class InsertionSort {
    public static void insertionSort(int[] input) {
        for (int i = 1; i < input.length; i++) {
            int j = i - 1;
            int temp = input[i];
            for (; j >= 0 && input[j] > temp; j--) {
                input[j + 1] = input[j];
            }
            input[j + 1] = temp;
        }
    }

    public static void main(String[] args) {
        int[] input = {  1, 5, 4, 2, 3  };
        insertionSort(input);
        for (int j : input) {
            System.out.print(j + " ");
        }
    }
}
```
I would suggest you add this code because this looks beginner-friendly and a lot of students doesn't get the object-oriented concepts, please add some comment if you, otherwise it is fine.